Setup a storybook

jdlrobson opened the pull request:
I'd like to migrate the documentation dfor our existing legacy components in
https://github.com/internetarchive/openlibrary/wiki/Design-Pattern-Library to the actual code repository
and begin documenting any new Vue components that are being added (It's been difficult to make sense of the
new Vue code so I think having documented examples will help).

Allowing exploration of the UI components inside storybook is bound
to make development a lot easier, particularly for people coming
back to the codebase after a long absence (like me!)

We've been using storybook in Wikimedia Foundation for all our major
projects and it's been a huge help.

To make our LESS files compatible with both storybook and the less
bundler, a few modifications to imports are required, but these should
have no impact on the resulting CSS

<!-- What issue does this PR close? -->
Closes #

<!-- What does this PR achieve? [feature|hotfix|fix|refactor] -->

### Technical
<!-- What should be noted about the implementation? -->

### Testing
<!-- Steps for reviewer to reproduce/verify what this PR does/fixes. -->
Run `npm run storybook` and explore.

### Screenshot
<!-- If this PR touches UI, please post evidence (screenshots) of it behaving correctly. -->

### Stakeholders
<!-- @ tag stakeholders of this bug -->
@cdrini @mekarpeles 

I'm happy to own this to migrate all of https://github.com/internetarchive/openlibrary/wiki/Design-Pattern-Library here. I think having legacy components alongside new Vue components will help highlight overlap and opportunities as well as flag existing problems with how the  CSS is organized.

cdrini commented on stories/Introduction.stories.mdx:
```diff
@@ -0,0 +1,14 @@
+<Meta title="Introduction" />
```
Could we also add a `README.md` file to `stories/` that shows what commands need to be run? Similar to https://github.com/internetarchive/openlibrary/blob/master/openlibrary/components/README.md ? The main questions I think would be nice to answer there would be:

- What's the flow I should use when developing a component? Does storybook have a live-reloading thing I can use, or is it more for documenting?
- What command(s) do I have to run to add a new story?
- What command(s) do I have to run to view the component "directory"?
- A link to a good storybook resource for the uninitiated would be nice :)

jdlrobson replied:
Storybook is primarily documentation for designers. Stuff like commands to run it are better done in the README.md

This is essentially the home page when you load the storybook which describes what's available and you can see it if you run `npm run storybook` and yes it does have hot reloading.

My work colleague wrote a good summary of how useful storybook is here:

The stories are essentially examples of components taken in isolation.

https://medium.com/freely-sharing-the-sum-of-all-knowledge/storypops-or-the-benefits-of-component-isolation-with-storybook-936746a6c1a

cdrini commented on package.json:
```diff
@@ -17,13 +17,19 @@
-    "test:js": "jest --testRegex tests/unit/**/test.*.js"
+    "test:js": "jest --testRegex tests/unit/**/test.*.js",
+    "storybook": "start-storybook -s . -p 6006",
+    "build-storybook": "build-storybook -s ."
```
When do I have to run the build step?

jdlrobson replied:
When I write components for other projects I run `npm run storybook` and it will boot up an instance using the stories defined in the stories folder.

It hot reloads so usually this is the primary way of building coordinates. The assumption being that if it works well in storybook it can be integrated into the site that's using them.

The build-storybook command creates a static folder that can be uploaded somewhere.

Here's one we use for Wikipedia mobile:
https://doc.wikimedia.org/MobileFrontend/master/js/ui/?path=/story/anchor--normal

Happy to demo all this at some point in 2021 if that's helpful.

cdrini commented on stories/FlatBookCover.stories.js:
```diff
@@ -0,0 +1,35 @@
+import FlatBookCover from '../openlibrary/components/LibraryExplorer/components/FlatBookCover.vue';
+
+export default {
+  title: 'Explorer/FlatBookCover',
```
```suggestion
  title: 'LibraryExplorer/FlatBookCover',
```

cdrini commented on stories/FlatBookCover.stories.js:
```diff
@@ -0,0 +1,35 @@
+const Template = (args, { argTypes }) => ({
+  props: Object.keys(argTypes),
+  components: { FlatBookCover },
+  template:'<flat-book-cover :book="book" />',
```
Can we use Pascal case? That's what I've been using in Vue land
```suggestion
  template:'<FlatBookCover :book="book" />',
```

jdlrobson replied:
I was curious about the use of camel case in the project. We've been using kebab case in Wikimedia projects  after weighing up eh trade offs [1, 2]

("Unfortunately, due to HTML’s case insensitivity, DOM templates must still use kebab-case."). Also note issues on case insensitive 

Might I suggest you add eslint supprot to enforce this if you've settled on that convention.
https://eslint.vuejs.org/rules/component-definition-name-casing.html

[1] https://vuejs.org/v2/style-guide/#Component-name-casing-in-templates-strongly-recommended
[2] https://vuejs.org/v2/guide/components-props.html#Prop-Casing-camelCase-vs-kebab-case

cdrini reviewed (commented):
Oh this looks exciting! I tried to add storybook while I was working on the Library Explorer, but couldn't quite get it up and running.

Do you have an idea of the flow you foresee us using storybook with? I have a rough idea of how it works/what it does, but haven't used it on another project before.

(Also heads up: We hit Travis' new limit for OS projects :/ So you might want to run linting locally to catch any errors).

(Also note: Here's where library explorer is on prod: https://openlibrary.org/explore )

mekarpeles commented:
I love this idea, thank you @jdlrobson

jdlrobson commented:
In case you want to run with this and merge, I've rebased and dropped FlatBookCover.stories for now as still waiting to understand the kebab case convention.

Once this patch is merged which adds the scaffolding I'll submit a new patchset for Vue and start documenting the existing legacy components as well as update https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide to explain how to get this up and running (which is linked to from Readme.md) and link to a netlify instance.

If you don't have the bandwidth to take on review for this right now, that's okay. Just let me know.

jdlrobson commented:
Lemme know you want to do a 1-on-1 @cdrini to get up to speed with storybook. Happy to knowledge share here.

jimchamp commented:
@jdlrobson I'm having some trouble getting this to run locally.  Should a `.storybook` configuration directory be included in this PR?

jdlrobson commented:
I'm afraid I don't remember it was almost a year ago :( 

It is possible something like https://github.com/wikimedia/mediawiki-skins-MinervaNeue/blob/master/.storybook/config.js is m issing.


If this patch is now of interest I'm happy to pull it up and work that out, but I don't want to do that if there's no timely plan to review it.

jimchamp commented:
I'll happily review this!  I think that Storybook will be really helpful.

jdlrobson commented:
Okay rebased and fleshed out the missing config! Still bare bones though... but should be enough to start documenting what's there and talking in a common language. I Don't know enough about the Vue stuff to incorporate that.

jimchamp reviewed (approved):
Lgtm!

jimchamp commented:
Thanks for the quick update!  I was able to get this up and running locally with no issues.  I think that having our components documented like this will be a great help.

jdlrobson commented:
As a next step it would be great to get the storybook outputing to github pages:
https://github.com/storybookjs/storybook-deployer

We can then link to it from https://github.com/internetarchive/openlibrary/wiki/Design-Pattern-Library

jimchamp commented:
That's a great idea!  Just created an issue for this, and plan on working on it next week.
